### PR TITLE
applets/swkbd: Software Keyboard Implementation

### LINKED ISSUE #1583: Games that still hangs while OS Software Keyboard should appear
**Zoo Resort 3D**

zoo 2

zoo

[LOG](http://pastebin.com/JdtzJ35y)

**Harvest Moon 3D A New Beginning**

reference [video](https://www.youtube.com/watch?v=i5XbZHzOQxc&list=PLMfPU_BxnPBaiCUeaxyf-r-cQG1auezxv)

harvest

[LOG](http://pastebin.com/8XuzqeVb)

### LINKED ISSUE #2095: Gundam Try Age SP hanging

### LINKED ISSUE #3346: Yu-Gi-Oh! Zexal - World Duel Carnival (3DS0994) hang at "Please enter your name"
The game runs smoothly until "enter your name", it's suddenly hang

My specs:

i5-6200U

Intel(R) HD Graphics 520

Nvidia 930M

8GB RAM

Citra Nightly HEAD-9b19c21

[LOGS.txt](https://github.com/citra-emu/citra/files/1599712/LOGS.txt)

### LINKED ISSUE #3382: My Pet Puppy 3D (USA) cannot input name
using the latest nightly
cannot input your name
related to #1583
puppy
[log.zip](https://github.com/citra-emu/citra/files/1630270/log.zip)

### LINKED ISSUE #3477: Can we just use System Username instead of hardcoded string?
Basically, couldn't you just use the System Username set in configuration instead of `Citra` that is hardcoded in https://github.com/citra-emu/citra/blob/master/src/core/hle/applets/swkbd.cpp#L80 ?

You know, until the new software keyboard is ready?

### PR DESCRIPTION
**Pull request by zhaowenlan1779**
Wow, so many additions.
Hmm, this is yet another swkbd implementation.

<del>But, this also includes a frontend applet interface by @jroweboy. For this part I only fixed typos.</del> Now removed, but core/frontend/swkbd part is mostly written by @jroweboy.
Also added log classes: `Applet` and `Applet_SWKBD`.
I guess the code is fairly easy to understand, and I won't have to explain my designs, so I'm just putting a screenshot here:
image

Buttons, custom texts, validations are properly implemented.

Fixes #1583, fix #2095, fix #3382, fix #3346

fix #3477 by last commit

---
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/3850)

### CODE REVIEW

File: src/core/hle/applets/swkbd.cpp [hunk 1]
```diff
@@ -69,22 +69,51 @@ ResultCode SoftwareKeyboard::StartImpl(Service::APT::AppletStartupParameter cons
...
+                       static_cast<int>(config.num_buttons_m1));
+        UNREACHABLE();
+    }
 
-    // TODO(Subv): Ask for input and write it to the shared memory
-    // TODO(Subv): Find out what are the possible values for the return code,
-    // some games seem to check for a hardcoded 2
-    config.return_code = 2;
-    config.text_length = 6;
+    config.text_length = static_cast<u16>(text.size() + 1);
```
Comment (wwylele) [inline src/core/hle/applets/swkbd.cpp]: IIRC I have tested that `text_length` doesn't include the terminating `'\0'`, and the length 6 in the code was a mistake.

File: src/core/hle/applets/swkbd.cpp [hunk 2]
```diff
@@ -114,5 +143,42 @@ void SoftwareKeyboard::Finalize() {
...
+
+Frontend::KeyboardConfig SoftwareKeyboard::ToFrontendConfig(SoftwareKeyboardConfig config) {
+    using namespace Frontend;
+    KeyboardConfig frontend_config;
+    frontend_config.button_config = static_cast<ButtonConfig>(config.num_buttons_m1);
+    frontend_config.accept_mode = static_cast<AcceptedInput>(config.valid_input);
+    frontend_config.multiline_mode = config.multiline;
+    frontend_config.max_text_length = config.max_text_length;
+    frontend_config.max_digits = config.max_digits;
+    std::wstring_convert<std::codecvt_utf8_utf16<char16_t>, char16_t> convert;
```
Comment (wwylele) [inline src/core/hle/applets/swkbd.cpp]: Use `Common::UTF16ToUTF8` and related functions, because the standard library (used to) have problems. We want to unify the codecvt code and update them according to the standard library status all together

File: src/core/hle/applets/swkbd.cpp [hunk 3]
```diff
@@ -114,5 +143,42 @@ void SoftwareKeyboard::Finalize() {
...
+    frontend_config.hint_text =
+        convert.to_bytes(reinterpret_cast<const char16_t*>(config.hint_text.data()));
```
Comment (wwylele) [inline src/core/hle/applets/swkbd.cpp]: Avoid `reinterpret_cast` and use `std::memcpy` to a buffer instead.

File: src/core/frontend/applets/swkbd.h [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+    MaxLengthExceeded,
+    BlankInputNotAllowed,
+    EmptyInputNotAllowed,
+};
+
+class SoftwareKeyboard : public AppletInterface {
+public:
+    explicit SoftwareKeyboard() : AppletInterface() {}
+    void Setup(const AppletConfig* config) override {
+        this->config = KeyboardConfig(*static_cast<const KeyboardConfig*>(config));
```
Comment (wwylele) [inline src/core/frontend/applets/swkbd.h]: This `static_cast` probably reveals a potential design problem. First of all it is obviously not type-safe, just like the old-school `void*`. Actually you would expect that every applet type would have to `static_cast` this because they want different configuration types => You actually want to make different `Setup` signature for different applet types => Then you probably want applets of different types to be, well, completely unrelated types/classes. This indeed would break your homogeneous container `registered_applets`, however it is not really necessary as you can totally just have `SoftwareKeyboard registered_softwarekeybord; MiiSelector registered_miiselector; ...` (And probably only these two applets worth GUI implementation)

In summary: the base class `AppletInterface` is really unnecessary as
 - It makes you do more unsafe casting, and makes it awkward to do anything related to a specific applet type.
 - It doesn't provide useful implementation to be shared with all applets
 - You don't benefit anything from polymorphism, as the polymorphic dispatch already happened in `core/hle/service/applet_manager -> core/hle/applet`

File: src/core/frontend/applets/swkbd.h [hunk 5]
```diff
@@ -0,0 +1,155 @@
+// Copyright 2018 Citra Emulator Project
+// Licensed under GPLv2 or any later version
+// Refer to the license.txt file included.
+
+#pragma once
+
+#include <codecvt>
+#include <locale>
```
Comment (wwylele) [inline src/core/frontend/applets/swkbd.h]: Where are these two used?

**Comment by zhaowenlan1779** [comment]
@neobrain I tested that and found the app just happily displayed a `?`

File: src/citra_qt/main.cpp [hunk 6]
```diff
@@ -1456,6 +1458,10 @@ int main(int argc, char* argv[]) {
     Camera::RegisterFactory("qt", std::make_unique<Camera::QtMultimediaCameraFactory>());
     Camera::QtMultimediaCameraHandler::Init();
 
+    // Register frontend applets
+    Frontend::RegisterDefaultApplets();
+    Frontend::RegisterSoftwareKeyboard(std::make_shared<QtKeyboard>(&main_window));
```
Comment (Subv) [inline src/citra_qt/main.cpp]: Use a reference if you're going to take the address of a local variable

File: src/core/frontend/applets/swkbd.h [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+    NotEmptyAndNotBlank, /// Empty or blank inputs (consisting solely of whitespace) are not
+                         /// accepted.
+    NotBlank,    /// Blank inputs (consisting solely of whitespace) are not accepted, but empty
+                 /// inputs are.
+    FixedLength, /// The input must have a fixed length (specified by maxTextLength in
+                 /// swkbdInit).
+};
+
+enum class ButtonConfig {
+    Single = 0, /// Ok button
```
Comment (lioncash) [inline src/core/frontend/applets/swkbd.h]: `= 0` can be omitted here.

File: src/core/frontend/applets/swkbd.h [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+
+#include <unordered_map>
+#include <utility>
+#include <vector>
+#include "common/assert.h"
+
+namespace Frontend {
+
+enum class AcceptedInput {
+    Anything = 0,        /// All inputs are accepted.
```
Comment (lioncash) [inline src/core/frontend/applets/swkbd.h]: `= 0` can be omitted here.

File: src/core/frontend/applets/swkbd.h [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+    Dual,       /// Cancel | Ok buttons
+    Triple,     /// Cancel | I Forgot | Ok buttons
+    None,       /// No button (returned by swkbdInputText in special cases)
+};
+
+/// Default English button text mappings. Frontends may need to copy this to internationalize it.
+static const char* BUTTON_OKAY = "Ok";
+static const char* BUTTON_CANCEL = "Cancel";
+static const char* BUTTON_FORGOT = "I Forgot";
+static const std::unordered_map<ButtonConfig, std::vector<std::string>> DEFAULT_BUTTON_MAPPING = {
```
Comment (lioncash) [inline src/core/frontend/applets/swkbd.h]: Note that every translation unit that makes use of this will have this whole map copied into it, bloating code size. This also applies to the strings.

File: src/core/frontend/applets/swkbd.h [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+    std::vector<std::string> button_text; /// Contains the button text that the caller provides
+    struct Filters {
+        bool prevent_digit;     /// Disallow the use of more than a certain number of digits
+                                /// TODO: how many is a certain number
+        bool prevent_at;        /// Disallow the use of the @ sign.
+        bool prevent_percent;   /// Disallow the use of the % sign.
+        bool prevent_backslash; /// Disallow the use of the \ sign.
+        bool prevent_profanity; /// Disallow profanity using Nintendo's profanity filter.
+        bool enable_callback;   /// Use a callback in order to check the input.
+    } filters;
```
Comment (lioncash) [inline src/core/frontend/applets/swkbd.h]: `Filter filters;` Prefer keeping structure instances separate from their definition.

File: src/core/frontend/applets/swkbd.h [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+     */
+    ValidationError Finalize(const std::string& text, u8 button);
+
+    KeyboardConfig config;
+    KeyboardData data;
+};
+
+class DefaultCitraKeyboard final : public SoftwareKeyboard {
+public:
+    void Setup(const KeyboardConfig* config) override {
```
Comment (lioncash) [inline src/core/frontend/applets/swkbd.h]: This should likely go in the cpp file as well, so no one has to rebuild everything including this header if changes are necessary to the implementation.

File: src/core/frontend/applets/swkbd.cpp [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+        // TODO(jroweboy): What does hardware do in this case?
+        NGLOG_CRITICAL(Frontend, "Application requested unknown validation method. Method: {}",
+                       static_cast<u32>(config.accept_mode));
+        UNREACHABLE();
+    }
+
+    return ValidationError::None;
+}
+
+ValidationError SoftwareKeyboard::ValidateButton(u8 button) {
```
Comment (lioncash) [inline src/core/frontend/applets/swkbd.cpp]: This can be a const member function

File: src/core/frontend/applets/swkbd.cpp [hunk 13]
```diff
@@ -0,0 +1,143 @@
...
+        LOG_INFO(Frontend, "App requested swkbd profanity filter, but its not implemented.");
+    }
+    if (config.filters.enable_callback) {
+        // TODO: check the callback
+        LOG_INFO(Frontend, "App requested a swkbd callback, but its not implemented.");
+    }
+    return ValidationError::None;
+}
+
+ValidationError SoftwareKeyboard::ValidateInput(const std::string& input) {
```
Comment (lioncash) [inline src/core/frontend/applets/swkbd.cpp]: This can be a const member function

File: src/core/frontend/applets/swkbd.cpp [hunk 14]
```diff
@@ -0,0 +1,143 @@
...
+// Refer to the license.txt file included.
+
+#include <cctype>
+#include "common/assert.h"
+#include "common/logging/log.h"
+#include "core/frontend/applets/swkbd.h"
+
+namespace Frontend {
+
+ValidationError SoftwareKeyboard::ValidateFilters(const std::string& input) {
```
Comment (lioncash) [inline src/core/frontend/applets/swkbd.cpp]: This can be a const member function.

File: src/core/frontend/applets/swkbd.h [hunk 15]
```diff
@@ -0,0 +1,150 @@
...
+     */
+    ValidationError ValidateButton(u8 button);
+
+    /**
+     * Runs all validation phases. If successful, stores the data so that the HLE applet in core can
+     * send this to the calling application
+     */
+    ValidationError Finalize(const std::string& text, u8 button);
+
+    KeyboardConfig config;
```
Comment (lioncash) [inline src/core/frontend/applets/swkbd.h]: These should likely be private.

File: src/core/hle/applets/swkbd.h [hunk 16]
```diff
@@ -15,33 +16,156 @@
...
+/// Maximum button text length, in UTF-16 code units.
+constexpr int MAX_BUTTON_TEXT_LEN = 16;
+/// Maximum hint text length, in UTF-16 code units.
+constexpr int MAX_HINT_TEXT_LEN = 64;
+/// Maximum filter callback error message length, in UTF-16 code units.
+constexpr int MAX_CALLBACK_MSG_LEN = 256;
+
+/// Keyboard types
+enum class SoftwareKeyboardType : u32 {
+    NORMAL = 0, ///< Normal keyboard with several pages (QWERTY/accents/symbol/mobile)
```
Comment (lioncash) [inline src/core/hle/applets/swkbd.h]: These should follow `TitleCase` if they're enum class entries. e.g.

```cpp
Normal,
QWERTY,
NumPad,
Western
```

`= 0` can also be omitted here.

File: src/core/hle/applets/swkbd.h [hunk 17]
```diff
@@ -15,33 +16,156 @@
...
+    QWERTY,     ///< QWERTY keyboard only.
+    NUMPAD,     ///< Number pad.
+    WESTERN,    ///< On JPN systems, a text keyboard without Japanese input capabilities,
+                /// otherwise same as SWKBD_TYPE_NORMAL.
+};
 
-    u16 max_text_length; ///< Maximum length of the input text
+/// Keyboard dialog buttons.
+enum class SoftwareKeyboardButtonConfig : u32 {
+    SINGLE_BUTTON = 0, ///< Ok button
```
Comment (lioncash) [inline src/core/hle/applets/swkbd.h]: Ditto, `= 0` can also be omitted here.

File: src/core/hle/applets/swkbd.h [hunk 18]
```diff
@@ -15,33 +16,156 @@
...
+    DUAL_BUTTON,       ///< Cancel | Ok buttons
+    TRIPLE_BUTTON,     ///< Cancel | I Forgot | Ok buttons
+    NO_BUTTON,         ///< No button (returned by swkbdInputText in special cases)
+};
 
-    INSERT_PADDING_BYTES(0x6E);
+/// Accepted input types.
+enum class SoftwareKeyboardValidInput : u32 {
+    ANYTHING = 0,      ///< All inputs are accepted.
```
Comment (lioncash) [inline src/core/hle/applets/swkbd.h]: Ditto. `= 0` can also be omitted here.

File: src/core/hle/applets/swkbd.h [hunk 19]
```diff
@@ -15,33 +16,156 @@
...
+    NOTBLANK, ///< Blank inputs (consisting solely of whitespace) are not accepted, but empty
+              /// inputs are.
+    FIXEDLEN, ///< The input must have a fixed length (specified by maxTextLength in
+              /// swkbdInit).
+};
 
-    char16_t display_text[65]; ///< Text to display when asking the user for input
+/// Keyboard password modes.
+enum class SoftwareKeyboardPasswordMode : u32 {
+    NONE = 0,   ///< Characters are not concealed.
```
Comment (lioncash) [inline src/core/hle/applets/swkbd.h]: Ditto. `= 0` can also be omitted here.

File: src/core/hle/applets/swkbd.h [hunk 20]
```diff
@@ -15,33 +16,156 @@
...
+    ALLOW_RESET = 1 << 6, ///< Allow the usage of a software-reset combination.
+    ALLOW_POWER = 1 << 7, ///< Allow the usage of the POWER button.
+    DEFAULT_QWERTY = 1 << 9, ///< Default to the QWERTY page when the keyboard is shown.
+};
+} // namespace SoftwareKeyboardFeature
 
-    u32 default_text_offset; ///< Offset of the default text in the output SharedMemory
+/// Keyboard filter callback return values.
+enum class SoftwareKeyboardCallbackResult : u32 {
+    OK = 0,   ///< Specifies that the input is valid.
```
Comment (lioncash) [inline src/core/hle/applets/swkbd.h]: Ditto. `= 0` can also be omitted here.

File: src/core/hle/applets/swkbd.h [hunk 21]
```diff
@@ -15,33 +16,156 @@
...
+    CLOSE,    ///< Displays an error message, then closes the keyboard.
+    CONTINUE, ///< Displays an error message and continues displaying the keyboard.
+};
 
-    INSERT_PADDING_WORDS(0x3);
+/// Keyboard return values.
+enum class SoftwareKeyboardResult : s32 {
+    NONE = -1,          ///< Dummy/unused.
```
Comment (lioncash) [inline src/core/hle/applets/swkbd.h]: Ditto

**Comment by zhaowenlan1779** [comment]
Rebased commits.

File: src/core/hle/applets/swkbd.cpp [hunk 22]
```diff
@@ -114,5 +143,44 @@ void SoftwareKeyboard::Finalize() {
...
+    frontend_config.accept_mode = static_cast<AcceptedInput>(config.valid_input);
+    frontend_config.multiline_mode = config.multiline;
+    frontend_config.max_text_length = config.max_text_length;
+    frontend_config.max_digits = config.max_digits;
+    std::u16string buffer;
+    buffer.resize(config.hint_text.size());
+    std::memcpy(buffer.data(), config.hint_text.data(), config.hint_text.size() * sizeof(u16));
+    frontend_config.hint_text = Common::UTF16ToUTF8(buffer);
+    frontend_config.has_custom_button_text =
+        !std::all_of(config.button_text.begin(), config.button_text.end(),
```
Comment (lioncash) [inline src/core/hle/applets/swkbd.cpp]: `<algorithm>` needs to be included.

File: src/core/hle/applets/swkbd.cpp [hunk 23]
```diff
@@ -114,5 +143,44 @@ void SoftwareKeyboard::Finalize() {
...
+Frontend::KeyboardConfig SoftwareKeyboard::ToFrontendConfig(SoftwareKeyboardConfig config) {
+    using namespace Frontend;
+    KeyboardConfig frontend_config;
+    frontend_config.button_config = static_cast<ButtonConfig>(config.num_buttons_m1);
+    frontend_config.accept_mode = static_cast<AcceptedInput>(config.valid_input);
+    frontend_config.multiline_mode = config.multiline;
+    frontend_config.max_text_length = config.max_text_length;
+    frontend_config.max_digits = config.max_digits;
+    std::u16string buffer;
+    buffer.resize(config.hint_text.size());
```
Comment (lioncash) [inline src/core/hle/applets/swkbd.cpp]: `std::utf16string buffer(config.hint_text.size(), 0);`

The resize isn't necessary.

File: src/core/frontend/applets/swkbd.cpp [hunk 24]
```diff
@@ -0,0 +1,143 @@
...
+    if ((error = ValidateInput(text)) != ValidationError::None) {
+        return error;
+    }
+    if ((error = ValidateButton(button)) != ValidationError::None) {
+        return error;
+    }
+    data = {text, button};
+}
+
+std::shared_ptr<SoftwareKeyboard> registered_swkbd;
```
[the remark this answers is not in the document]
Comment (lioncash) [reply src/core/frontend/applets/swkbd.cpp]: I disagree entirely.

1. It's not static so it has external linkage (that still doesn't resolve other issues).
2. It makes it a pain to launch more than one instance of the core or make library portion of the emulator a proper library in the future. I don't think making the core worse in this regard is a good idea (other globals existing should not be seen as an encouragement to do the same).
3. Just because something else is haphazard doesn't mean other things should also be that haphazard as well.
4. Even if it's file-scope, it's still functionally a global from a behavioral point of view.

[the remark this answers is not in the document]
Comment (lioncash) [reply src/core/frontend/applets/swkbd.cpp]: Fundamentally if an instance of something is providing behavior that the system provides, it belongs in the System class, full-stop.

File: src/core/hle/applets/swkbd.cpp [hunk 25]
```diff
@@ -93,13 +123,7 @@ void SoftwareKeyboard::Update() {
...
 
 void SoftwareKeyboard::DrawScreenKeyboard() {
-    auto bottom_screen = Service::GSP::GetFrameBufferInfo(0, 1);
-    auto info = bottom_screen->framebuffer_info[bottom_screen->index];
-
-    // TODO(Subv): Draw the HLE keyboard, for now just zero-fill the framebuffer
-    Memory::ZeroBlock(info.address_left, info.stride * 320);
-
-    Service::GSP::SetBufferSwap(1, info);
+    // TODO(Subv): Draw the HLE keyboard, for now just do nothing
```
Comment (zhaowenlan1779) [inline src/core/hle/applets/swkbd.cpp]: Explanation: The implementation of the function is useless, for all cases now:
1. You don't need to wait for `DefaultCitraKeyboard`, so you will hardly see the effect
1. `QtKeyboard` blocks before returning from Setup, so frame display will not be updated until input dialog has finished. The black screen will still be displayed, but it is a splash after input dialog finishes, which looks quite stupid.

**Comment by zhaowenlan1779** [comment]
@lioncash Moved registered_swkbd to System

File: src/core/hle/applets/swkbd.h [hunk 26]
```diff
@@ -71,6 +195,8 @@ class SoftwareKeyboard final : public Applet {
     void Finalize();
 
 private:
+    Frontend::KeyboardConfig ToFrontendConfig(SoftwareKeyboardConfig config);
```
Comment (lioncash) [inline src/core/hle/applets/swkbd.h]: `config` should likely be a const reference, otherwise this does quite a large copy. The member function can likely be a const member function as well

File: src/core/frontend/applets/swkbd.h [hunk 27]
```diff
@@ -0,0 +1,131 @@
...
+
+enum class ButtonConfig {
+    Single, /// Ok button
+    Dual,   /// Cancel | Ok buttons
+    Triple, /// Cancel | I Forgot | Ok buttons
+    None,   /// No button (returned by swkbdInputText in special cases)
+};
+
+/// Default English button text mappings. Frontends may need to copy this to internationalize it.
+static const char* BUTTON_OKAY = "Ok";
```
Comment (lioncash) [inline src/core/frontend/applets/swkbd.h]: These can just be:

`constexpr char name[] = "Text"`

const file-scope variables are static by default. Using arrays instead of pointers marginally saves executable space (a pointer address doesn't need to be stored in the executable, only the string data itself, so it's basically free).

File: src/citra_qt/applets/swkbd.cpp [hunk 28]
```diff
@@ -0,0 +1,123 @@
...
+
+#include <QDialogButtonBox>
+#include <QLabel>
+#include <QLineEdit>
+#include <QMessageBox>
+#include <QString>
+#include <QVBoxLayout>
+#include "citra_qt/applets/swkbd.h"
+
+QtKeyboardValidator::QtKeyboardValidator(QtKeyboard* keyboard) : keyboard(keyboard) {}
```
Comment (lioncash) [inline src/citra_qt/applets/swkbd.cpp]: This will cause a variable shadowing warning. The parameter should be `keyboard_` or something else

File: src/citra_qt/applets/swkbd.cpp [hunk 29]
```diff
@@ -0,0 +1,123 @@
...
+
+QtKeyboardValidator::State QtKeyboardValidator::validate(QString& input, int& pos) const {
+    if (keyboard->ValidateFilters(input.toStdString()) == Frontend::ValidationError::None) {
+        return State::Acceptable;
+    } else {
+        return State::Invalid;
+    }
+}
+
+QtKeyboardDialog::QtKeyboardDialog(QWidget* parent, QtKeyboard* keyboard) : keyboard(keyboard) {
```
Comment (lioncash) [inline src/citra_qt/applets/swkbd.cpp]: Ditto about keyboard

File: src/citra_qt/applets/swkbd.cpp [hunk 30]
```diff
@@ -0,0 +1,123 @@
...
+        {ValidationError::MaxLengthExceeded,
+         tr("Text is too long (should be no more than %1 characters)")
+             .arg(keyboard->config.max_text_length)},
+        {ValidationError::BlankInputNotAllowed, tr("Blank input is not allowed")},
+        {ValidationError::EmptyInputNotAllowed, tr("Empty input is not allowed")},
+    };
+    QMessageBox::critical(this, tr("Validation error"), VALIDATION_ERROR_MESSAGES.at(error));
+}
+
+QtKeyboard::QtKeyboard(QWidget& parent) : parent(parent) {}
```
Comment (lioncash) [inline src/citra_qt/applets/swkbd.cpp]: Ditto about variable shadowing

File: src/core/core.cpp [hunk 31]
```diff
@@ -199,6 +199,10 @@ const Service::SM::ServiceManager& System::ServiceManager() const {
     return *service_manager;
 }
 
+void System::RegisterSoftwareKeyboard(std::shared_ptr<Frontend::SoftwareKeyboard> swkbd) {
+    registered_swkbd = swkbd;
```
Comment (lioncash) [inline src/core/core.cpp]: This can be `registered_swkbd = std::move(swkbd);` to avoid an unnecessary atomic reference count increment and decrement

File: src/core/hle/applets/swkbd.cpp (hunk 25 above)
Comment (zhaowenlan1779) [reply src/core/hle/applets/swkbd.cpp]: @jroweboy 
Hmm. That sounds correct.
Perhaps I will make it not block on Setup and use wait condition or something.

**Comment by BreadFish64** [comment]
is it possible to limit the number of characters in the textbox, instead of showing a popup when there's too many characters
